app/bundles/EmailBundle/Swiftmailer/Momentum/DTO/TransmissionDTO/RecipientDTO.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+namespace Mautic\EmailBundle\Swiftmailer\Momentum\DTO\TransmissionDTO;
+
+use Mautic\EmailBundle\Swiftmailer\Momentum\DTO\TransmissionDTO\RecipientDTO\AddressDTO;
+
+/**
+ * Class RecipientDTO.
+ */
+final class RecipientDTO implements \JsonSerializable
+{
+    /**
+     * @var string|null
+     */
+    private $returnPath = null;
+
+    /**
+     * @var AddressDTO
+     */
+    private $address;
+
+    /**
+     * @var array
+     */
+    private $tags = [];
+
+    /**
+     * @var array
+     */
+    private $metadata = [];
+
+    /**
+     * @var array
+     */
+    private $substitutionData = [];
+
+    /**
+     * RecipientDTO constructor.
+     *
+     * @param AddressDTO $addressDTO
+     * @param array      $metadata
+     * @param array      $substitutionData
+     */
+    public function __construct(AddressDTO $addressDTO, $metadata = [], $substitutionData = [])
+    {
+        $this->address          = $addressDTO;
+        $this->metadata         = $metadata;
+        $this->substitutionData = $substitutionData;
+    }
+
+    /**
+     * @param null|string $returnPath
+     *
+     * @return RecipientDTO
+     */
+    public function setReturnPath($returnPath)
```

Comment:
Many of this setter functions in all DTOs are never used. What is their purpose?
It is better to build object via constructor and avoid setter if possible.

Reply:
Actually those DTOs are wrong. They shouldn't accept any parameter in __construct and have everything optional. That is because DTO should act as container for data so it doesn't matter if it's valid from some perspective or not.

If that would be entity, than yes - it should be in constructor as much as possible and from that perspective our current entities are wrong. But we need both - entity and DTO working together each on different site.

Actually with this DTO approach you can use just DTOs for communication between Model and rest of application so framework doesn't even need to know, if you are using ORM or NotORM - it never saw entity.